### ISSUE TITLE
9.0 stage2 builds exit after building razor

### ISSUE THREAD

**Issue by tmds** [dotnet/source-build#4608]
The 9.0 stage2 builds we have in our internal CI terminate after the razor build finished successfully:

These are the last lines that are printed.

```
  [20:55:43.27] Building razor...done
  New artifact(s) after building razor:
    -> Microsoft.AspNetCore.Mvc.Razor.Extensions.Tooling.Internal/9.0.0-preview.24463.3
    -> Microsoft.CodeAnalysis.Razor.Tooling.Internal/9.0.0-preview.24463.3
    -> Microsoft.NET.Sdk.Razor.SourceGenerators.Transport/9.0.0-preview.24463.3
```

There is no error reported after those, the build script just returns.

Builds from last week (Sep 10) don't show the issue.

cc @MichaelSimons @omajid

**Comment by omajid** [comment]
If I am looking at the same build, there were some error messages in the log, long before the build failed:

```
  dotnet/src/fsharp/eng/SourceBuild.props(43,5): error MSB3073: The command "./build.sh --bootstrap --skipBuild -bl --tfm net9.0 /p:SourceBuildUseMonoRuntime=false" exited with code -1. [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]

  Build FAILED.

  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(26,36): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_codeopt has been constrained to be type 'Code'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(26,46): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_Macros has been constrained to be type 'Macro list'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(26,55): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_Rules has been constrained to be type 'Rule list'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(36,26): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_macro has been constrained to be type 'Macro'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(41,31): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_regexp has been constrained to be type 'Regexp'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(46,26): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_rule has been constrained to be type 'Rule'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(52,30): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_args has been constrained to be type 'RuleArgument list'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(52,33): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clauses has been constrained to be type 'Clause list'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fs(411,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optbar has been constrained to be type 'unit'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(65,44): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clause has been constrained to be type 'Clause'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fsy(84,78): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_charset has been constrained to be type ''a -> 'b'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fslex/fslexpars.fs(260,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype__startspec has been constrained to be type 'obj'. [dotnet/src/fsharp/buildtools/fslex/fslex.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(25,59): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_decls has been constrained to be type '(ParserSpec -> ParserSpec) list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(25,71): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_headeropt has been constrained to be type 'Code'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(25,134): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_rules has been constrained to be type '(Identifier * Rule list) list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(35,39): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_decl has been constrained to be type 'ParserSpec -> ParserSpec'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(38,107): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_idents has been constrained to be type 'Identifier list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(46,42): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_rule has been constrained to be type 'Identifier * Rule list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(47,69): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clauses has been constrained to be type 'Rule list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(411,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optbar has been constrained to be type 'unit'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(431,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optsemi has been constrained to be type 'unit'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(50,51): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_clause has been constrained to be type 'Rule'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,55): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_syms has been constrained to be type 'Identifier list'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fsy(51,58): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype_optprec has been constrained to be type 'Identifier option'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/buildtools/fsyacc/fsyaccpars.fs(223,20): error FS0064: This construct causes code to be less generic than indicated by the type annotations. The type variable 'gentype__startspec has been constrained to be type 'obj'. [dotnet/src/fsharp/buildtools/fsyacc/fsyacc.fsproj] [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
  dotnet/src/fsharp/eng/SourceBuild.props(43,5): error MSB3073: The command "./build.sh --bootstrap --skipBuild -bl --tfm net9.0 /p:SourceBuildUseMonoRuntime=false" exited with code -1. [dotnet/artifacts/source-built-sdks/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj]
      0 Warning(s)
      26 Error(s)
```

**Comment by tmds** [comment]
Looks like a dup of https://github.com/dotnet/source-build/issues/4603.

**Comment by tmds** [comment]
@MichaelSimons why does the vmr build continue to build several repos after the fsharp repo failed to build?

**Comment by mthalman** [comment]
> @MichaelSimons why does the vmr build continue to build several repos after the fsharp repo failed to build?

I'm guessing that parallel builds are enabled and `RunEachTargetSeparately` is set to `false`. If so, that would cause all parallel projects in the process of being built to wait until completion, even if one of them failed. More context is here: https://github.com/dotnet/installer/pull/18824. Do you happen to have `ContinuousIntegrationBuild` set to `true`?

**Comment by tmds** [comment]
> Do you happen to have ContinuousIntegrationBuild set to true?

We're not passing the `--ci` argument or anything `ContinuousIntegrationBuild` related.

**Comment by MichaelSimons** [comment]
> > Do you happen to have ContinuousIntegrationBuild set to true?
> 
> We're not passing the `--ci` argument or anything `ContinuousIntegrationBuild` related.

Could you enable the binlog and share the Build.binlog?  TIA.

**Comment by MichaelSimons** [comment]
Friendly ping @tmds - Can you get me a binlog.

**Comment by tmds** [comment]
The only thing we're passing to the vmr build script is `--source-only`. I assume that that should not cause the build to be parallel.

I'm going to close this issue as the fsharp fix went to 9.0.

I will have a look into the parallel behavior myself at a later time, and open up an issue when I have something to share.
